### DISCUSSION TITLE
Recommendations for rebooting container host

### DISCUSSION THREAD

**Discussion by superjamie** [nextcloud/all-in-one#5185] [Questions] [answered]
Say I need to reboot my container host running AIO (eg: for OS updates).

Should I log into the AIO interface and manually stop all the containers, then stop the master container on the commandline, then reboot?

Or is it safe to just reboot the system and let Docker pull the containers down and automatically restart them all on reboot?

**Comment by szaimen** [comment, marked as answer]
Hi,

> Or is it safe to just reboot the system and let Docker pull the containers down and automatically restart them all on reboot?

Yes, this is the idea.

**Reply by baptit** [reply]
When I shut down nextcloud-aio-mastercontainer (using compose, docker rootless), all other nextcloud containners remain up. I assume this is not the idea. What is going wrong?

**Reply by gggeek** [reply]
You have to first tell the MC to stop the other containers, I guess.

Whether that should happen automatically or not when stopping the MC is a design choice, and one that atm is, imho, not clearly documented.

I wonder if you could add dependencies (but not the full containers defs) between the containers in your Compose file, to make sure that all the other ones stop before the MC...

**Reply by nihebe** [reply]
I think there is not just a lack of clear documentation but a functional deficit – whenever I shutdown the host system or the docker daemon **without** stopping the non-MC containers through AIO interface something goes wrong. Mostly, the Nextcloud container doesn't start again after and I have to rebuild it.

I tried to automate the AIO-shutdown with system.d (cf [few comments below](https://github.com/nextcloud/all-in-one/discussions/5185#discussioncomment-16839466)), but this doesn't work either. So the dependency approach via yaml sounds interesting. Any ideas on how to accomplish this? Seems difficult, as the yaml just defines the MC and "doesn't know" the other containers, which are instantiated by the MC.

**Reply by gggeek** [reply]
> Seems difficult, as the yaml just defines the MC and "doesn't know" the other containers, which are instantiated by the MC

Indeed it would be more a hack than anything else - even if it works :-D

I think that most users would expect the whole stack to be shut down cleanly and automatically when the MC is stopped, contrary to what happens atm, but maybe that precludes some interesting scenarios, such as hot-upgrading the MC without disruption to the NC application?

In any case, having a concise one-liner command to stop both the MC and the dependent containers is clearly a need felt by many users...

**Reply by gggeek** [reply]
> that precludes some interesting scenarios, such as hot-upgrading the MC without disruption to the NC application?

Thinking more about this, it should be doable to achieve the best of both worlds with the following workflow:
- by default, when stopping, the MC stops the other containers before self
- when needing to do a self-upgrade or similar stop operation, the MC sets a variable into itself (have yet to figure out if that would be best handled on persistent volume disk or only on local container disk, probably the latter), which tells its shutdown script not to stop the other contaniers

@szaimen wdyt?

**Reply by szaimen** [reply]
I dont think we should go this way. The docker daemon should handle the stopping for us correctly. In any way, feel free to open a feature request for this here: https://github.com/nextcloud/all-in-one/discussions/new?category=ideas

**Reply by baptit** [reply]
See request at https://github.com/nextcloud/all-in-one/discussions/8321. It is a different question as to whether docker daemon should wait for all (user) containers to gracefully stop when it is terminated. I guess we need both approaches.

**Comment by superjamie** [comment]
Many thanks!

**Reply by baptit** [reply]
@szaimen
Hi,
Running Nextcloud-AIO as rootless docker, it seems as if pulling down AIO does not trigger stopping the Nextcloud containers. They continue to run:
```
~> docker container ls
CONTAINER ID   IMAGE                                                  COMMAND                  CREATED              STATUS                        PORTS                                                                     NAMES
85470c482cd7   ghcr.io/nextcloud-releases/aio-apache:latest           "/start.sh /usr/bin/…"   About a minute ago   Up About a minute (healthy)   80/tcp, 0.0.0.0:11000->11000/tcp                                          nextcloud-aio-apache
b437ebb90f5d   ghcr.io/nextcloud-releases/aio-nextcloud:latest        "/start.sh /usr/bin/…"   About a minute ago   Up About a minute (healthy)   9000/tcp                                                                  nextcloud-aio-nextcloud
644a4baafd0d   ghcr.io/nextcloud-releases/aio-imaginary:latest        "/start.sh"              About a minute ago   Up About a minute (healthy)                                                                             nextcloud-aio-imaginary
41fed0b0971d   ghcr.io/nextcloud-releases/aio-fulltextsearch:latest   "/bin/tini -- /usr/l…"   2 minutes ago        Up About a minute (healthy)   9200/tcp, 9300/tcp                                                        nextcloud-aio-fulltextsearch
f6b9e8f4e224   ghcr.io/nextcloud-releases/aio-clamav:latest           "/start.sh /usr/bin/…"   2 minutes ago        Up 2 minutes (healthy)                                                                                  nextcloud-aio-clamav
674eeba3fb60   ghcr.io/nextcloud-releases/aio-redis:latest            "/start.sh"              2 minutes ago        Up 2 minutes (healthy)        6379/tcp                                                                  nextcloud-aio-redis
a662be8b5fc4   ghcr.io/nextcloud-releases/aio-postgresql:latest       "/start.sh"              2 minutes ago        Up 2 minutes (healthy)        5432/tcp                                                                  nextcloud-aio-database
8e15c4e8a07b   ghcr.io/nextcloud-releases/aio-whiteboard:latest       "/start.sh"              2 minutes ago        Up 2 minutes (healthy)        3002/tcp                                                                  nextcloud-aio-whiteboard
f8e94893fdd2   ghcr.io/nextcloud-releases/aio-notify-push:latest      "/start.sh"              2 minutes ago        Up 2 minutes (healthy)                                                                                  nextcloud-aio-notify-push
912e409d533f   ghcr.io/nextcloud-releases/aio-collabora:latest        "/start-collabora-on…"   2 minutes ago        Up 2 minutes (healthy)        9980/tcp                                                                  nextcloud-aio-collabora
b42ed33a0859   nextcloud/all-in-one:latest                            "/start.sh"              7 minutes ago        Up 3 minutes (healthy)        80/tcp, 8443/tcp, 9000/tcp, 0.0.0.0:8080->8080/tcp, [::]:8080->8080/tcp   nextcloud-aio-mastercontainer
646fe00b0424   nickfedor/watchtower                                   "/watchtower"            5 days ago           Up 5 days (healthy)           8080/tcp                                                                  watchtower
~> docker compose -f ~/nextcloud/compose.yaml down
[+] down 2/2
 ✔ Container nextcloud-aio-mastercontainer Removed                                                                                                                                                                                                            3.2s
 ✔ Network nextcloud_default               Removed                                                                                                                                                                                                            0.1s
~> docker container ls
CONTAINER ID   IMAGE                                                  COMMAND                  CREATED         STATUS                   PORTS                              NAMES
85470c482cd7   ghcr.io/nextcloud-releases/aio-apache:latest           "/start.sh /usr/bin/…"   2 minutes ago   Up 2 minutes (healthy)   80/tcp, 0.0.0.0:11000->11000/tcp   nextcloud-aio-apache
b437ebb90f5d   ghcr.io/nextcloud-releases/aio-nextcloud:latest        "/start.sh /usr/bin/…"   2 minutes ago   Up 2 minutes (healthy)   9000/tcp                           nextcloud-aio-nextcloud
644a4baafd0d   ghcr.io/nextcloud-releases/aio-imaginary:latest        "/start.sh"              2 minutes ago   Up 2 minutes (healthy)                                      nextcloud-aio-imaginary
41fed0b0971d   ghcr.io/nextcloud-releases/aio-fulltextsearch:latest   "/bin/tini -- /usr/l…"   2 minutes ago   Up 2 minutes (healthy)   9200/tcp, 9300/tcp                 nextcloud-aio-fulltextsearch
f6b9e8f4e224   ghcr.io/nextcloud-releases/aio-clamav:latest           "/start.sh /usr/bin/…"   2 minutes ago   Up 2 minutes (healthy)                                      nextcloud-aio-clamav
674eeba3fb60   ghcr.io/nextcloud-releases/aio-redis:latest            "/start.sh"              3 minutes ago   Up 3 minutes (healthy)   6379/tcp                           nextcloud-aio-redis
a662be8b5fc4   ghcr.io/nextcloud-releases/aio-postgresql:latest       "/start.sh"              3 minutes ago   Up 3 minutes (healthy)   5432/tcp                           nextcloud-aio-database
8e15c4e8a07b   ghcr.io/nextcloud-releases/aio-whiteboard:latest       "/start.sh"              3 minutes ago   Up 3 minutes (healthy)   3002/tcp                           nextcloud-aio-whiteboard
f8e94893fdd2   ghcr.io/nextcloud-releases/aio-notify-push:latest      "/start.sh"              3 minutes ago   Up 3 minutes (healthy)                                      nextcloud-aio-notify-push
912e409d533f   ghcr.io/nextcloud-releases/aio-collabora:latest        "/start-collabora-on…"   3 minutes ago   Up 3 minutes (healthy)   9980/tcp                           nextcloud-aio-collabora
646fe00b0424   nickfedor/watchtower                                   "/watchtower"            5 days ago      Up 5 days (healthy)      8080/tcp                           watchtower
```
Is this due to some misconfiguration on my side? I am looking for a simple, automated way to do maintanence on the host OS which would require a safe method to stop Nextcloud beforehand.

**Reply by szaimen** [reply]
Hi, see https://github.com/nextcloud/all-in-one#how-to-stopstartupdate-containers-or-trigger-the-daily-backup-from-a-script-externally

**Comment by baptit** [comment]
So is running this script better than to just reboot the system and let Docker pull the containers down and automatically restart them all on reboot?

**Reply by nihebe** [reply]
That's what I understood so far.
It would be very desirable to have a solution for automating this – did anyone already find a way to build a Debian systemd-service that takes care of calling the shutdown script when a shutdown is initiated?
I'm hosting NC AIO in a Debian VM and need this for example when the hypervisor shuts down all VMs in case of an energy failure (UPS time is limited).

**Reply by nihebe** [reply]
I'm trying it like this now… 

`sudo nano /etc/systemd/system/nextcloud-aio-shutdown.service`
```
[Unit]
Description=Gracefully stop Nextcloud AIO containers
DefaultDependencies=no
Before=shutdown.target reboot.target halt.target
After=docker.service network.target
Requires=docker.service

[Service]
Type=oneshot
ExecStart=/usr/bin/docker exec nextcloud-aio-mastercontainer sudo -u www-data php /var/www/docker-aio/php/src/Cron/StopContainers.php
TimeoutStartSec=25
RemainAfterExit=true

[Install]
WantedBy=halt.target reboot.target shutdown.target
```
Test with: `sudo systemctl start nextcloud-aio-shutdown.service`
--> Works perfectly, all containers except AIO Mastercontainer shut down.

Then I enabled it:
`sudo systemctl daemon-reload`
`sudo systemctl enable nextcloud-aio-shutdown.service`

But when I shutdown, it does not seem like the shutdown script is executed. At least `sudo journalctl -u nextcloud-aio-shutdown.service` does not show any entries from the shutdown point in time and also NC AIO does not start correctly (which is typical for the Nextcloud-container after killing it).

So I went back to my snapshot… but now the question is how to trigger the script correctly. Any ideas?

**Reply by szaimen** [reply]
Hi, you might want to have a look at https://github.com/nextcloud/all-in-one#how-to-stopstartupdate-containers-or-trigger-the-daily-backup-from-a-script-externally

**Reply by nihebe** [reply]
Hi, 
thanks, saw that one earlier.
But I think it makes no difference if I execute `/usr/bin/docker exec nextcloud-aio-mastercontainer sudo -u www-data php /var/www/docker-aio/php/src/Cron/StopContainers.php` or `sudo docker exec --env STOP_CONTAINERS=1 nextcloud-aio-mastercontainer /daily-backup.sh` in this case.
It seems to be a Debian/systemd-issue, don't you think so?

**Reply by szaimen** [reply]
For starting you need to run the script on boot with START_CONTAINERS=1

**Reply by nihebe** [reply]
My aim is to **stop** AIO containers when the OS shutdown (i.e. OS receives shutdown signal from Hypervisor).
And I think people in this issue/thread are having difficulties in configuring a scripted event for Linux host systems that calls AIO container shutdown.
One could argue that this is a question for a Debian forum – but maybe someone in the NC community has a good recommendation here. :)

**Reply by szaimen** [reply]
Hm... usually containers should be automatically stopped together with the host OS by the docker daemon...

**Reply by nihebe** [reply]
Unfortunately, this does not work reliably – docker daemon gives containers 10 sec to shutdown, which seems to be insufficient for Nextcloud AIO. Thus, the Nextcloud container is stopped ungracefully and must be recreated after restart.

**Reply by szaimen** [reply]
I just created anothed guide. Can you check it out? 
- see https://github.com/nextcloud/all-in-one/discussions/8324
